### DISCUSSION TITLE
Jagged Array Initialization Like Java

### DISCUSSION THREAD

**Discussion by Timothy-Liu** [dotnet/csharplang#7154] [Language Ideas]
Sincerely hope C# supports jagged array initialization like Java.

For example, ```int[][] matrix = new int[3][4];```

Thanks!

**Comment by bernd5** [comment]
Why not using a multi dimensional array?

**Comment by FaustVX** [comment]
May be this will be supported with the inline-array/fixed-buffer proposal #7130 #1314

**Comment by HaloFour** [comment]
I think it's important to note that jagged array initializers in Java are shorthand for a loop that creates each subarray:

```java
// int[][] matrix = new int[3][4];

// allocates four separate arrays in the heap
int[][] matrix = new int[3][];
for (int i = 0; i < 3; i++) {
    matrix[i] = new int[4];
}
```

Given that .NET and C# has always supported multidimensional arrays, which are much more efficient to initialize and to index, I think it was intentional that jagged array initialization was not supported:

```cs
int[,] matrix = new int[3,4]; // allocates space for 12 contiguous ints
```

That does seem like it might change with inline arrays, which will retain the efficiency of multi-dimensional arrays by emitting anonymous inline array types:

```cs
int[3][4] matrix = new int[3][4];
// is really
Buffer4<int>[] matrix = new Buffer4<int>[3]; // allocates space for 3 contiguous 4-int sized value types
```

Which I assume will have compatibility concerns and can't be passed to a method that expects `int[][]`?

**Reply by CyrusNajmabadi** [reply]
> Which I assume will have compatibility concerns and can't be passed to a method that expects int[][]?

Correct

**Reply by FaustVX** [reply]
It's weird that `int[3][4]` is not compatible with `int[][]`.
I understand why, it's not the same underlying type, but confusing.

**Reply by TahirAhmadov** [reply]
Easy enough to create an extension method `.ToArray()`, right? Can `int[3][4]` be passed with `in`?

**Reply by FaustVX** [reply]
@TahirAhmadov 
Yes of course, but that would create a new object, so more memory and time to copy each element.
And each operation on the 'callee' array will not be reflected to the 'caller' array.
```C#
int[3][4] a = new int[3](4];
M(a.ToArray())
// a[0][0] hasn't changed

void M(int[][] b) => b[0] = new int[4]; // this is valid today, so this won't change
```

**Reply by TahirAhmadov** [reply]
@FaustVX yes but there is no solution to this, it's by design... hence my question - can you pass `int[3][4]` by `ref/in` to a method which expects it? Then you can create a new overload for stack arrays.
Also, you can always do `int[][] arr = a.ToArray(); M(arr); arr.CopyTo(a);` - yes, clunky, but if `M` is a third-party method, it's the best option...

**Reply by HaloFour** [reply]
@TahirAhmadov 

> Easy enough to create an extension method

I'm not sure that's true.  Every arity of the array is a different type, so on what type would you define that extension method?  I don't really see anything in the proposal that would suggest language support that would make this feasible.  Although, it seems you could always write extension methods for `ReadOnlySpan<T>` and use the emitted `AsReadOnlySpan()` method on the `BufferX<T>` type.

> Then you can create a new overload for stack arrays.

That's assuming you own the code which exposes the method that accepts `int[][]`, and given the code for the two really can't be shared you'd basically be copypastaing the whole thing.  From the proposal it also doesn't seem like you can declare a method that accepts `int[3][4]`, you might have to explicitly declare it as `BufferX<T>`?  Not sure, though.
